File: ts-sdk/src/storage.ts
import { parseBlobResponse, parseSupythonResponse, wrapNetworkError } from './errors';
import type {
  Bucket,
  CreateBucketOptions,
  CreateSignedUrlOptions,
  FileBody,
  SignedUrl,
  StorageObject,
  UploadOptions,
} from './types/storage';
import type { GetHeaders, SupythonResponse } from './errors';

/**
 * StorageClient manages buckets and produces {@link StorageBucket} handles
 * for object operations. It delegates all HTTP to `fetch` and reads headers
 * from the supplied `getHeaders` callback per-request.
 */
export class StorageClient {
  /** @internal — exposed for StorageBucket only; not part of the public API. */
  readonly _url: string;
  /** @internal */
  readonly _getHeaders: GetHeaders;

  constructor(url: string, getHeaders: GetHeaders) {
    this._url = url.replace(/\/$/, '');
    this._getHeaders = getHeaders;
  }

  /** Return a {@link StorageBucket} handle for the named bucket. */
  from(bucketName: string): StorageBucket {
    return new StorageBucket(this, bucketName);
  }

  /** Create a new bucket. */
  async createBucket(name: string, options?: CreateBucketOptions): Promise<SupythonResponse<Bucket>> {
    const body: Record<string, unknown> = { name, public: options?.public ?? false };
    if (options?.file_size_limit !== undefined) body.file_size_limit = options.file_size_limit;
    if (options?.allowed_mime_types !== undefined) body.allowed_mime_types = options.allowed_mime_types;
    return this._request<Bucket>('/bucket', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });
  }

  /** List all buckets. */
  async listBuckets(): Promise<SupythonResponse<Bucket[]>> {
    return this._request<Bucket[]>('/bucket', { method: 'GET' });
  }

  /** Get metadata for a single bucket by name. */
  async getBucket(name: string): Promise<SupythonResponse<Bucket>> {
    return this._request<Bucket>(`/bucket/${name}`, { method: 'GET' });
  }

  /** Delete a bucket. */
  async deleteBucket(name: string): Promise<SupythonResponse<null>> {
    return this._request<null>(`/bucket/${name}`, { method: 'DELETE' }, true);
  }

  private async _request<T>(
    path: string,
    init: RequestInit,
    expectStatus204 = false,
  ): Promise<SupythonResponse<T>> {
    return wrapNetworkError(async () => {
      const resp = await fetch(`${this._url}${path}`, {
        ...init,
        headers: { ...this._getHeaders(), ...(init.headers as Record<string, string> | undefined) },
      });
      if (expectStatus204 && resp.status === 204) {
        return { data: null as unknown as T, error: null };
      }
      return parseSupythonResponse<T>(resp, 'storage');
    }, 'storage');
  }
}

/**
 * StorageBucket performs object-level operations inside a single bucket.
 *
 * Object keys (`path`) are passed through verbatim — callers are responsible
 * for sanitisation. Keys must not start with a leading `/`.
 */
export class StorageBucket {
  private readonly _client: StorageClient;
  private readonly _bucket: string;

  constructor(client: StorageClient, bucket: string) {
    this._client = client;
    this._bucket = bucket;
  }

  /**
   * Upload an object to `path` inside this bucket.
   *
   * The body is sent as `multipart/form-data` under field name `file`.
   * `Blob`/`File` bodies are appended directly; other values are wrapped
   * in a `Blob` with `options.contentType` (default `application/octet-stream`).
   */
  async upload(path: string, body: FileBody, options?: UploadOptions): Promise<SupythonResponse<StorageObject>> {
    return wrapNetworkError(async () => {
      const form = new FormData();
      if (body instanceof Blob) {
        form.append('file', body);
      } else {
        const blob = new Blob([body as BlobPart], { type: options?.contentType ?? 'application/octet-stream' });
        form.append('file', blob, path.split('/').pop() ?? 'file');
      }
      const resp = await fetch(`${this._client._url}/object/${this._bucket}/${path}`, {
        method: 'POST',
        headers: this._client._getHeaders(),
        body: form,
      });
      return parseSupythonResponse<StorageObject>(resp, 'storage');
    }, 'storage');
  }

  /** Download the object at `path` as a {@link Blob}. */
  async download(path: string): Promise<SupythonResponse<Blob>> {
    return wrapNetworkError(async () => {
      const resp = await fetch(`${this._client._url}/object/${this._bucket}/${path}`, {
        headers: this._client._getHeaders(),
      });
      return parseBlobResponse(resp);
    }, 'storage');
  }

  /** Remove a single object at `path`. Batch removal is not yet supported. */
  async remove(path: string): Promise<SupythonResponse<null>> {
    return wrapNetworkError(async () => {
      const resp = await fetch(`${this._client._url}/object/${this._bucket}/${path}`, {
        method: 'DELETE',
        headers: this._client._getHeaders(),
      });
      if (resp.status === 204) return { data: null, error: null };
      return parseSupythonResponse<null>(resp, 'storage');
    }, 'storage');
  }

  /**
   * Create a time-limited signed URL for the object at `path`.
   *
   * @param expiresIn — lifetime in seconds; omit to use the server default.
   */
  async createSignedUrl(path: string, options?: CreateSignedUrlOptions): Promise<SupythonResponse<SignedUrl>> {
    const body = options?.expiresIn !== undefined
      ? JSON.stringify({ expires_in: options.expiresIn })
      : JSON.stringify({});
    return wrapNetworkError(async () => {
      const resp = await fetch(`${this._client._url}/object/sign/${this._bucket}/${path}`, {
        method: 'POST',
        headers: { ...this._client._getHeaders(), 'Content-Type': 'application/json' },
        body,
      });
      return parseSupythonResponse<SignedUrl>(resp, 'storage');
    }, 'storage');
  }

  /**
   * Return the public URL for an object. No HTTP call is made.
   *
   * Note: this does not verify that the bucket is actually public —
   * callers must ensure the bucket settings match their use-case.
   */
  getPublicUrl(path: string): string {
    return `${this._client._url}/object/public/${this._bucket}/${path}`;
  }
}

File: src/supython/client/_storage.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import httpx

from ._auth import SupythonResponse
from ._config import _parse_error_detail


@dataclass
class StorageError:
    code: str
    message: str
    status: int


@dataclass
class BucketResponse:
    id: str
    name: str
    owner: str | None
    public: bool
    file_size_limit: int | None
    allowed_mime_types: list[str] | None
    created_at: str
    updated_at: str


@dataclass
class ObjectResponse:
    id: str
    bucket_id: str
    bucket: str
    name: str
    owner: str
    size: int
    mime_type: str | None
    etag: str | None
    created_at: str
    updated_at: str


@dataclass
class SignedUrlResponse:
    signed_url: str
    token: str
    expires_at: str
    expires_in: int


def _make_storage_error(resp: httpx.Response) -> StorageError:
    try:
        body = resp.json()
    except Exception:
        msg = resp.text or f"HTTP {resp.status_code}"
        return StorageError("network_error", msg, resp.status_code)
    code, message = _parse_error_detail(body)
    return StorageError(code, message, resp.status_code)


def _parse_bucket(body: dict[str, Any]) -> BucketResponse:
    return BucketResponse(
        id=str(body["id"]),
        name=body["name"],
        owner=str(body["owner"]) if body.get("owner") else None,
        public=body["public"],
        file_size_limit=body.get("file_size_limit"),
        allowed_mime_types=body.get("allowed_mime_types"),
        created_at=body["created_at"],
        updated_at=body["updated_at"],
    )


def _parse_object(body: dict[str, Any]) -> ObjectResponse:
    return ObjectResponse(
        id=str(body["id"]),
        bucket_id=str(body["bucket_id"]),
        bucket=body["bucket"],
        name=body["name"],
        owner=str(body["owner"]),
        size=body["size"],
        mime_type=body.get("mime_type"),
        etag=body.get("etag"),
        created_at=body["created_at"],
        updated_at=body["updated_at"],
    )


class StorageBucket:
    def __init__(self, client: StorageClient, bucket_name: str) -> None:
        self._client = client
        self._bucket_name = bucket_name

    def _headers(self) -> dict[str, str]:
        return self._client._headers()

    async def upload(
        self,
        path: str,
        data: bytes,
        *,
        content_type: str | None = None,
    ) -> SupythonResponse[ObjectResponse]:
        url = f"{self._client._url}/object/{self._bucket_name}/{path}"
        files = {"file": (path, data, content_type or "application/octet-stream")}
        try:
            resp = await self._client._http.post(
                url, files=files, headers=self._headers()
            )
        except httpx.HTTPError as exc:
            return SupythonResponse(error=StorageError("network_error", str(exc), 0))

        if resp.status_code >= 400:
            return SupythonResponse(error=_make_storage_error(resp))

        return SupythonResponse(data=_parse_object(resp.json()))

    async def download(self, path: str) -> SupythonResponse[bytes]:
        url = f"{self._client._url}/object/{self._bucket_name}/{path}"
        try:
            resp = await self._client._http.get(url, headers=self._headers())
        except httpx.HTTPError as exc:
            return SupythonResponse(error=StorageError("network_error", str(exc), 0))

        if resp.status_code >= 400:
            return SupythonResponse(error=_make_storage_error(resp))

        return SupythonResponse(data=resp.content)

    async def remove(self, path: str) -> SupythonResponse[None]:
        url = f"{self._client._url}/object/{self._bucket_name}/{path}"
        try:
            resp = await self._client._http.delete(url, headers=self._headers())
        except httpx.HTTPError as exc:
            return SupythonResponse(error=StorageError("network_error", str(exc), 0))

        if resp.status_code >= 400:
            return SupythonResponse(error=_make_storage_error(resp))

        return SupythonResponse(data=None)

    async def create_signed_url(
        self, path: str, *, expires_in: int | None = None
    ) -> SupythonResponse[SignedUrlResponse]:
        url = f"{self._client._url}/object/sign/{self._bucket_name}/{path}"
        body: dict[str, Any] = {}
        if expires_in is not None:
            body["expires_in"] = expires_in
        try:
            resp = await self._client._http.post(
                url, json=body or None, headers=self._headers()
            )
        except httpx.HTTPError as exc:
            return SupythonResponse(error=StorageError("network_error", str(exc), 0))

        if resp.status_code >= 400:
            return SupythonResponse(error=_make_storage_error(resp))

        data = resp.json()
        return SupythonResponse(
            data=SignedUrlResponse(
                signed_url=data["signed_url"],
                token=data["token"],
                expires_at=data["expires_at"],
                expires_in=data["expires_in"],
            )
        )

    def get_public_url(self, path: str) -> str:
        return f"{self._client._url}/object/public/{self._bucket_name}/{path}"


class StorageClient:
    def __init__(self, base_url: str, anon_key: str, client: Any) -> None:
        self._url = base_url
        self._anon_key = anon_key
        self._client = client
        self._http = httpx.AsyncClient()

    def _headers(self) -> dict[str, str]:
        headers: dict[str, str] = {}
        if self._anon_key:
            headers["apikey"] = self._anon_key
        access_token = self._client._access_token
        if access_token:
            headers["Authorization"] = f"Bearer {access_token}"
        return headers

    def from_(self, bucket_name: str) -> StorageBucket:
        return StorageBucket(self, bucket_name)

    async def create_bucket(
        self,
        *,
        name: str,
        public: bool = False,
        file_size_limit: int | None = None,
        allowed_mime_types: list[str] | None = None,
    ) -> SupythonResponse[BucketResponse]:
        body: dict[str, Any] = {"name": name, "public": public}
        if file_size_limit is not None:
            body["file_size_limit"] = file_size_limit
        if allowed_mime_types is not None:
            body["allowed_mime_types"] = allowed_mime_types

        try:
            resp = await self._http.post(
                f"{self._url}/bucket", json=body, headers=self._headers()
            )
        except httpx.HTTPError as exc:
            return SupythonResponse(error=StorageError("network_error", str(exc), 0))

        if resp.status_code >= 400:
            return SupythonResponse(error=_make_storage_error(resp))

        return SupythonResponse(data=_parse_bucket(resp.json()))

    async def list_buckets(self) -> SupythonResponse[list[BucketResponse]]:
        try:
            resp = await self._http.get(f"{self._url}/bucket", headers=self._headers())
        except httpx.HTTPError as exc:
            return SupythonResponse(error=StorageError("network_error", str(exc), 0))

        if resp.status_code >= 400:
            return SupythonResponse(error=_make_storage_error(resp))

        return SupythonResponse(data=[_parse_bucket(b) for b in resp.json()])

    async def get_bucket(self, name: str) -> SupythonResponse[BucketResponse]:
        try:
            resp = await self._http.get(
                f"{self._url}/bucket/{name}", headers=self._headers()
            )
        except httpx.HTTPError as exc:
            return SupythonResponse(error=StorageError("network_error", str(exc), 0))

        if resp.status_code >= 400:
            return SupythonResponse(error=_make_storage_error(resp))

        return SupythonResponse(data=_parse_bucket(resp.json()))

    async def delete_bucket(self, name: str) -> SupythonResponse[None]:
        try:
            resp = await self._http.delete(
                f"{self._url}/bucket/{name}", headers=self._headers()
            )
        except httpx.HTTPError as exc:
            return SupythonResponse(error=StorageError("network_error", str(exc), 0))

        if resp.status_code >= 400:
            return SupythonResponse(error=_make_storage_error(resp))

        return SupythonResponse(data=None)
